### PR TITLE
chore: validate node network and disallow mainnet

### PR DESCRIPTION
**Pull request by Extheoisah**
closes https://github.com/bitcoin-dev-project/sim-ln/issues/29

### CODE REVIEW

File: sim-lib/src/lib.rs [hunk 1]
```diff
@@ -108,6 +109,7 @@ pub struct NodeInfo {
     pub pubkey: PublicKey,
     pub alias: String,
     pub features: NodeFeatures,
+    pub networks: Vec<String>,
```
Comment (okjodom) [inline sim-lib/src/lib.rs:112]: node info should have singular `network`

File: sim-lib/src/lib.rs [hunk 2]
```diff
@@ -275,6 +277,44 @@ impl Simulation {
         Ok(())
     }
 
+    // validate_node_network validates that the nodes that are on the same bitcoin network.
+    // e.g. node A and B are on testnet, node C is on mainnet should return an error.
+    async fn validate_node_network(&self) -> Result<(), LightningError> {
+        let mut networks = Vec::new();
+        let mainnet = Network::Bitcoin.to_string().to_lowercase();
+
+        for node in self.nodes.values() {
```
Comment (okjodom) [inline sim-lib/src/lib.rs:286]: if each node info declares just one `network`, maybe you could replace the validation logic below with something like

```rs

let mainnet = Network::Bitcoin.to_string().to_lowercase();
let mut valid_network = Option::None;

for node in self.nodes.values() {
    let network = node.lock().await.get_info().network;

    if network == mainnet {
        return Err(LightningError::ValidationError(
            "mainnet is not supported".to_string(),
        ));
    }

    let valid_replaced = valid_network.replace(network.clone());

    if valid_replaced.is_some() && valid_replaced != network {
        return Err(LightningError::ValidationError(format!(
            "nodes are not on the same network {}",
            network,
        )));
    }
}

Ok(())
```

File: sim-lib/src/lib.rs [hunk 3]
```diff
@@ -283,6 +323,7 @@ impl Simulation {
         }
 
         self.validate_activity().await?;
+        self.validate_node_network().await?;
```
Comment (okjodom) [inline sim-lib/src/lib.rs:326]: prefer we validate network before activities :)

File: sim-lib/src/lnd.rs [hunk 4]
```diff
@@ -66,6 +67,10 @@ impl LndNode {
                     .map_err(|err| LightningError::GetInfoError(err.to_string()))?,
                 features: parse_node_features(features.keys().cloned().collect()),
                 alias,
+                networks: chains
```
Comment (okjodom) [inline sim-lib/src/lnd.rs:70]: so CLN advertises just a single network while LND can advertise multiple? 🤦🏿

Comment (Extheoisah) [reply sim-lib/src/lnd.rs:70]: Yes. And its why i made it a vec

File: sim-lib/src/lib.rs [hunk 5]
```diff
@@ -326,7 +367,7 @@ impl Simulation {
         let mut success = true;
         while let Some(res) = tasks.join_next().await {
             if let Err(e) = res {
-                log::error!("Task exited with error: {e}.");
+                log::error!("Task exited with error: {}.", e);
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:370]: Does this need to change?

File: sim-lib/src/lnd.rs (hunk 4 above)
Comment (carlaKC) [reply sim-lib/src/lnd.rs:70]: FUN LIGHTNING IS FUN OKAY.

**Review by carlaKC — COMMENTED** [review]
🙌 welcome to sim-ln 🙌 

Can be [simplified](https://github.com/bitcoin-dev-project/sim-ln/pull/112/files#r1334503033) by just assuming one network, but looking good!

File: sim-lib/src/lib.rs (hunk 5 above)
Comment (Extheoisah) [reply sim-lib/src/lib.rs:366]: Not really. The syntax Just didn't look right to me.

**Comment by Extheoisah** [comment]
comments addressed! @carlaKC @okjodom

File: sim-lib/src/lib.rs [hunk 6]
```diff
@@ -275,13 +277,48 @@ impl Simulation {
...
+                        network,
+                    )));
+                }
+            } else {
+                valid_network = Some(network);
+            }
+        }
+
+        log::info!(
+            "Simulation is running on the network: {}",
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:307]: nit: end logs with a period `.`

File: sim-lib/src/lib.rs (hunk 5 above)
Comment (carlaKC) [reply sim-lib/src/lib.rs:366]: I think this is pretty canonical rust so let's leave as is?

File: sim-lib/src/lib.rs [hunk 7]
```diff
@@ -275,13 +277,48 @@ impl Simulation {
...
+            if let Some(valid) = &valid_network {
+                if valid != &network {
+                    return Err(LightningError::ValidationError(format!(
+                        "nodes are not on the same network {}",
+                        network,
+                    )));
+                }
+            } else {
+                valid_network = Some(network);
+            }
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:303]: We could shorten this a little with `valid_network = valid_network.unwrap_or(network.clone());` on 294, lets us get rid of the conditional / fill our own value if it's currently none.

File: sim-lib/src/lib.rs [hunk 8]
```diff
@@ -275,13 +277,48 @@ impl Simulation {
         Ok(())
     }
 
+    // validate_node_network validates that the nodes that are on the same bitcoin network.
+    // e.g. node A and B are on testnet, node C is on mainnet should return an error.
+    async fn validate_node_network(&self) -> Result<(), LightningError> {
+        let mut valid_network = Option::None;
+        let mainnet = Network::Bitcoin.to_string().to_lowercase();
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:284]: This string is "bitcoin" which won't match the `network` set by LND "mainnet" (there's divergence between LDK and LND in naming).

**Review by carlaKC — COMMENTED** [review]
Looking good!  Successfully tested 2/3 validation checks: 

Multiple chains: 
```
Error: Get info error: LND node is not connected to a single chain
```

Different networks:
```
Error: Lightning Error: ValidationError("nodes are not on the same network regtest")

Caused by:
    Config validation failed: nodes are not on the same network regtest
```

Disallowing mainnet needs an update for LND (we're checking `bitcoin` but LND sets `mainnet` as its network value). Will test CLN on next pass, but looks straightforward.

File: sim-lib/src/lib.rs (hunk 8 above)
Comment (Extheoisah) [reply sim-lib/src/lib.rs]: Yes, an oversight on my part. fixed here -> https://github.com/bitcoin-dev-project/sim-ln/pull/112/commits/5d2bfac7f1a9d20630899b10602bdcf8cae38b75

File: sim-lib/src/lib.rs [hunk 9]
```diff
@@ -275,13 +276,50 @@ impl Simulation {
...
+                return Err(LightningError::ValidationError(
+                    "mainnet is not supported".to_string(),
+                ));
+            }
+
+            valid_network = valid_network.take().or_else(|| Some(network.clone()));
+            if let Some(valid) = &valid_network {
+                if valid != &network {
+                    return Err(LightningError::ValidationError(format!(
+                        "nodes are not on the same network {}",
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:300]: nittynit: "not on the same network: {}"

File: sim-lib/src/lib.rs [hunk 10]
```diff
@@ -275,13 +276,50 @@ impl Simulation {
...
     }
 
+    // validate_node_network validates that the nodes that are on the same bitcoin network.
+    // e.g. node A and B are on testnet, node C is on mainnet should return an error.
+    async fn validate_node_network(&self) -> Result<(), LightningError> {
+        let mut valid_network = Option::None;
+        let unsupported_networks: HashSet<String> = vec!["mainnet", "bitcoin"]
+            .into_iter()
+            .map(|s| s.to_string())
+            .collect();
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:286]: Fewer LOC to just declare `mut HashSet` then `insert`, but no strong feelings.

File: sim-lib/src/lib.rs [hunk 11]
```diff
@@ -275,13 +276,50 @@ impl Simulation {
         Ok(())
     }
 
+    // validate_node_network validates that the nodes that are on the same bitcoin network.
+    // e.g. node A and B are on testnet, node C is on mainnet should return an error.
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:280]: Comment nitpick: it validates that the nodes are all on the same network and ensures that we're not running on mainnet.

**Review by carlaKC — COMMENTED** [review]
tACK!
```
Error: Lightning Error: ValidationError("mainnet is not supported")

Caused by:
    Config validation failed: mainnet is not supported
```

I think you can go ahead and squash this all into a single commit then we can merge. Left some last nits, but none of them are blocking.

**Review by carlaKC — APPROVED** [review]
LGTM!

**Review by okjodom — APPROVED** [review]

File: sim-lib/src/lnd.rs [hunk 12]
```diff
@@ -59,13 +60,20 @@ impl LndNode {
             .map_err(|err| LightningError::GetInfoError(err.to_string()))?
             .into_inner();
 
+        if chains.len() != 1 {
```
Comment (sr-gi) [inline sim-lib/src/lnd.rs:63]: nit: we can use `is_empty` here to be more idiomatic

File: sim-lib/src/lnd.rs [hunk 13]
```diff
@@ -59,13 +60,20 @@ impl LndNode {
...
+        }
+
         Ok(Self {
             client,
             info: NodeInfo {
                 pubkey: PublicKey::from_str(&identity_pubkey)
                     .map_err(|err| LightningError::GetInfoError(err.to_string()))?,
                 features: parse_node_features(features.keys().cloned().collect()),
                 alias,
+                network: chains[0].network.clone(),
```
Comment (sr-gi) [inline sim-lib/src/lnd.rs:76]: nit: we can use `chains.swap_remove(0).network` here to avoid unnecessarily copying a value.

**Review by sr-gi — COMMENTED** [review]
Post-merge review.

I'm looking at this given I'm trying to rebase for #103 and I just realized the way this is handled is troublesome :(

From #103 onwards, we are using `NodeInfo` to store information about all nodes in our config, not only the ones we control. However, we cannot query the network of those.

Furthermore, once we've checked the network the nodes we're controlling are running at, there is no need to store the `network` in `NodeInfo`, given we're not using that anymore. Hence, we should just add a method to query the `network` instead of making this part of `get_info`

**Pull request merged**